api/glossary_usefulconcepts.md

```diff
@@ -0,0 +1,156 @@
+
+## Decorators (in Python)
+These are used in programming as wrappers around a function or object in order to modify it’s behaviour. Decorators are placed above a function with following characteristic syntax:
+`@nameofdecorator`
+Each decorator will modify an aspect of the original function or a class’s behaviour in a specific way without permanently modifying it. The decorator returns a version of the object that was wrapped.
+One can have more than one decorator, and take the function to be modified as input.
+
+For example:
+```buildoutcfg
+@decorator_1
+@decorator_2
+def function_one():
+    pass
+
+function_temp = decorator_2(function_one)
+function_final = decorator_1(function_temp)
+```
+
+##### Example decorators found in Baobab
+[@marshall_with](https://flask-restplus.readthedocs.io/en/stable/marshalling.html)
+
+[@auth_required](https://flask-security-too.readthedocs.io/en/stable/patterns.html)
+
+These can all be easily Googled for further clarity and descriptions as to how they modify the respective functions.
+
+## Endpoints
+The [API endpoint](https://smartbear.com/learn/performance-monitoring/api-endpoints/) is the end of a communication channel. When there are two interacting APIs, these interact with each other at the endpoint. This is the location from which APIs can access the network resources they need to carry out their functions.
+APIs work using `requests` and `responses` in that they request information from a web application or web server and then receive a response. Endpoints send the requests and indicate where the resource lives. These are stored in the [routes.py](https://github.com/deep-learning-indaba/Baobab/blob/develop/api/app/routes.py) file in the [Baobab repository](https://github.com/deep-learning-indaba/Baobab).
+
+## Models
+These are an example of a Python object, specifically a database.
+
+## Relational databases
+The type of database we are interested in is called a relational database (and this is why ORM [Object Relational Mapping] is so relevant).
+Relational databases set up relationships between items efficiently, and each data are stored in a table. The rows contain the items and the columns contain the properties of the items. The ORM (in our case, SQLAlchemy) is written in a language that is part of the web framework language (in our case, Python), allowing us to communicate with the database that is more familiar than SQL.
+
+####Organisation of the data in a database versus a class:
+**Class:**
+
+|ClassName|
+|---------|
+|**id:** int|
+|**name:** str|
+|**descriptions:** str|
+
+**Relational database:**
+
+|ID:|Name:|Description:|
+|---|---|---|
+|INT|Char|VarChar|
+
+### Database keys
+Keys are fundamental in the creation of relationships between data (intra- and inter-tables). Further, they assist in maintaining uniqueness in a table; ensure consistent and valid data as well as allow for efficient data retrieval.
+
+|Candidate keys (multiple, one can be a primary key)||
```

Comment:
I'm a bit confused by this table - what are the rows and columns? Or were you just doing it  to get a 2 column layout? If that's the case then think it would be better to just put headings and bullet points :)

Reply:
I was trying to show the overlap between different keys, but Markdown is limited in terms of tables so it didn't work out as I had hoped. I've changed it to bullets and added full sentences :)